aten/src/ATen/cuda/CuSparseHandlePool.cpp

```diff
@@ -37,7 +37,7 @@ cusparseHandle_t getCurrentCUDASparseHandle() {
   AT_CUDA_CHECK(cudaGetDevice(&device));
 
   if (!myPoolWindow)
-    myPoolWindow.reset(pool.newPoolWindow());
+    myPoolWindow.reset(pool->newPoolWindow());
```

Comment:
I don't think `newPoolWindow` needs to exist.  We could replace this call (and the other `myPoolWindow.reset()` calls) with
```
myPoolWindow.reset(new decltype(pool)::element_type::PoolWindow(pool));
```
Then each `DeviceThreadHandlePool` wouldn't need to inherit from `enable_shared_from_this`.

Maybe `newPoolWindow` is a good layer of indirection though, because it gives each `DeviceThreadHandlePool` the option to customize `PoolWindow` creation.  I think this is a matter of style/preference and not blocking.

Reply:
Yeah, that makes sense.

The original code used `newPoolWindow` and I didn't want to change the structure too much.

Either way, I think removing `newPoolWindow` is best done in another PR because it isn't enabled by the change in this PR (i.e. `myPoolWindow.reset(new decltype(pool)::PoolWindow(pool));` would have worked before this PR).

Also, if `newPoolWindow` is removed, then it probably makes sense to move `PoolWindow` out from `DeviceThreadHandlePool` so it's no longer a nested class.

I think doing both of those here would make this PR a bit messier.

Reply:
Agreed, I think the PR is fine as is.  In my [original handle pool PR](https://github.com/pytorch/pytorch/pull/15668/files), targeting cudnn handles only, `DeviceThreadHandlePool` was not a thing.  Members of what is now `DeviceThreadHandlePool` (its `unordered_maps` and mutex) were standalone objects and `PoolWindow` was a standalone class.  When @zasdfgbnm extended the pooling logic to cover cublas and cusparse handles, he added `DeviceThreadHandlePool` with instantiations to organize the stuff for each pool.

Great catch!!